src/mpid/ch4/generic/am/mpidig_am_recv.h

```diff
@@ -139,16 +139,40 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_handle_unexpected(void *buf, MPI_Aint count,
     rreq->status.MPI_SOURCE = MPIDIG_REQUEST(rreq, rank);
     rreq->status.MPI_TAG = MPIDIG_REQUEST(rreq, tag);
 
-    mpi_errno = MPIDIG_copy_from_unexp_req(rreq, buf, datatype, count);
-    MPIR_ERR_CHECK(mpi_errno);
+    if (MPIDIG_REQUEST(rreq, buffer)) {
+        /* if we have an unexp buffer, we just need to copy the data in it to the user buffer */
+        mpi_errno = MPIDIG_copy_from_unexp_req(rreq, buf, datatype, count);
+        MPIR_ERR_CHECK(mpi_errno);
+        MPIDIG_REQUEST(rreq, req->status) &= ~MPIDIG_REQ_UNEXPECTED;
 
-    MPIDIG_REQUEST(rreq, req->status) &= ~MPIDIG_REQ_UNEXPECTED;
+        /* If this is a synchronous send, send back the reply indicating that the message has been
+         * matched. */
+        if (MPIDIG_REQUEST(rreq, req->status) & MPIDIG_REQ_PEER_SSEND) {
+            mpi_errno = MPIDIG_reply_ssend(rreq);
+            MPIR_ERR_CHECK(mpi_errno);
+        }
 
-    /* If this is a synchronous send, send back the reply indicating that the message has been
-     * matched. */
-    if (MPIDIG_REQUEST(rreq, req->status) & MPIDIG_REQ_PEER_SSEND) {
-        mpi_errno = MPIDIG_reply_ssend(rreq);
-        MPIR_ERR_CHECK(mpi_errno);
+    } else {
+        MPI_Aint unexp_data_sz = 0;
+        unexp_data_sz = MPIDIG_REQUEST(rreq, count);
```

Comment:
Add comment to note that `count` is the incoming data size that is set during target_msg_cb.

Reply:
Will do.